include/boost/geometry/algorithms/for_each.hpp

```diff
@@ -54,209 +56,310 @@ namespace detail { namespace for_each
+        bool result = f(p);
+        *it = p;
+        return result;
+    }
+};
 
-        // So we now loop manually.
 
-        for (typename boost::range_iterator<Range>::type
-                it = boost::begin(range); it != boost::end(range); ++it)
+struct fe_point_range
+{
+    template <typename Range, typename Functor>
+    static inline bool apply(Range& range, Functor&& f)
+    {
+        auto const end = boost::end(range);
+        for (auto it = boost::begin(range); it != end; ++it)
         {
-            f(*it);
+            if (! fe_point_call_f<Range>::apply(it, f))
+            {
+                return false;
+            }
         }
+
+        return true;
     }
 };
 
 
 template
 <
     typename Range,
-    typename Point = typename add_const_if_c
-        <
-            boost::is_const<Range>::value,
-            typename point_type<Range>::type
-        >::type,
-    bool UseReferences =
-        ( boost::is_const<Range>::value
-       || boost::is_same
-            <
-                typename boost::range_reference<Range>::type,
-                Point&
-            >::value )
+    bool UseReferences = fe_point_type_is_referencable<Range>::value
 >
-struct fe_range_per_segment_call_f
+struct fe_segment_call_f
 {
     template <typename Iterator, typename Functor>
-    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    static inline bool apply(Iterator it0, Iterator it1, Functor&& f)
     {
         // Implementation for real references (both const and mutable)
         // and const proxy references.
         // If const proxy references are returned by iterators
         // then const real references here prevents temporary
         // objects from being destroyed.
-        Point& p0 = *it0;
-        Point& p1 = *it1;
-        model::referring_segment<Point> s(p0, p1);
-        f(s);
+        typedef typename fe_point_type<Range>::type point_type;
+        point_type& p0 = *it0;
+        point_type& p1 = *it1;
+        model::referring_segment<point_type> s(p0, p1);
+        return f(s);
     }
 };
 
-template <typename Range, typename Point>
-struct fe_range_per_segment_call_f<Range, Point, false>
+template <typename Range>
+struct fe_segment_call_f<Range, false>
 {
     template <typename Iterator, typename Functor>
-    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    static inline bool apply(Iterator it0, Iterator it1, Functor&& f)
     {
         // Mutable proxy references returned by iterators.
         // Temporary points have to be created and assigned afterwards.
-        Point p0 = *it0;
-        Point p1 = *it1;
-        model::referring_segment<Point> s(p0, p1);
-        f(s);
+        typedef typename fe_point_type<Range>::type point_type;
+        point_type p0 = *it0;
+        point_type p1 = *it1;
+        model::referring_segment<point_type> s(p0, p1);
+        bool result = f(s);
         *it0 = p0;
         *it1 = p1;
+        return result;
```

Comment:
There is a reason why `f(s)` is called (and returned boolean cached) before `p0` and `p1` are accessed, right?

Shouldn't the assignments be skipped all together if the `f` fails?

```cpp
if (!f(s))
    return false;

*it0 = p0;
*it1 = p1;
return true;
```

Reply:
I agree with you in general however my intention was to generate exactly the same outcome as in case of the version for lvalue mutable iterators. The caller don't know whether or not the points are going to be modified inside `f`. The points can be modified and the functor can still return `false`. The final assignment would reflect that, i.e. this code would work exactly how it works for real references.

An example would be an algorithm transforming N points of a geometry or something and stopping as soon as they are transformed. I don't know if this algorithm makes sense but I'm just showing a scenario when the outcomes would be different.

Or am I missing something?

Reply:
> The points can be modified and the functor can still return false.

It is an important clarification.
It makes the difference to the approach which I assumed, i.e. in-out arguments remain untouched on failure.

My questions have been answered, thanks.